src/planner/parabolic/HauserParabolicSmootherHelpers.cpp

```diff
@@ -0,0 +1,216 @@
+      }
+      else
+      {
+        rampNd.blendAttempts++;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool doShortcut(
+    ParabolicRamp::DynamicPath& dynamicPath,
+    aikido::constraint::TestablePtr testable,
+    double timelimit,
+    double checkResolution,
+    double tolerance,
+    aikido::util::RNG& rng)
+{
+  if (timelimit < 0.0)
+    throw std::invalid_argument("Timelimit should be non-negative");
+  if (checkResolution <= 0.0)
+    throw std::invalid_argument("Check resolution should be positive");
+  if (tolerance < 0.0)
+    throw std::invalid_argument("Tolerance should be non-negative");
+
+  SmootherFeasibilityCheckerBase base(testable, checkResolution);
+  ParabolicRamp::RampFeasibilityChecker feasibilityChecker(&base, tolerance);
+
+  std::chrono::time_point<std::chrono::system_clock> startTime
+      = std::chrono::system_clock::now();
+  double elapsedTime = 0;
+
+  bool success = false;
+  while (elapsedTime < timelimit && dynamicPath.ramps.size() > 3)
+  {
+    std::uniform_real_distribution<> dist(0.0, dynamicPath.GetTotalTime());
+    double t1 = dist(rng);
+    double t2 = dist(rng);
+    if (dynamicPath.TryShortcut(t1, t2, feasibilityChecker))
+    {
+      success = true;
+    }
+
+    elapsedTime = std::chrono::duration_cast<std::chrono::duration<double>>(
+                      std::chrono::system_clock::now() - startTime)
+                      .count();
+  }
+  return success;
+}
+
+bool doBlend(
+    ParabolicRamp::DynamicPath& dynamicPath,
+    aikido::constraint::TestablePtr testable,
+    double blendRadius,
+    int blendIterations,
+    double checkResolution,
+    double tolerance)
+{
+  if (blendIterations <= 0)
+    throw std::invalid_argument("Blend iterations should be positive");
+  if (blendRadius <= 0.0)
+    throw std::invalid_argument("Blend radius should be positive");
+  if (checkResolution <= 0.0)
+    throw std::invalid_argument("Check resolution should be positive");
+  if (tolerance < 0.0)
+    throw std::invalid_argument("Tolerance should be non-negative");
+
+  SmootherFeasibilityCheckerBase base(testable, checkResolution);
+  ParabolicRamp::RampFeasibilityChecker feasibilityChecker(&base, tolerance);
+
+  // Mark all of the ramps in the initial trajectory as "original". We'll
+  // only try to blend transitions between these ramps.
+  for (unsigned int i = 0; i < dynamicPath.ramps.size(); ++i)
+  {
+    dynamicPath.ramps[i].blendAttempts = 0;
+  }
+
+  double dtShortcut = blendRadius;
+
+  // tryBlend always starts at the beginning of the trajectory.
+  // Without this bookkeeping, tryBlend would could any blend that fails
+  // multiple times.
+  // We continue to call tryBlend() in a loop until it finds nothing left
+  // to blend.
+  for (int attempt = 0; attempt < blendIterations; ++attempt)
+  {
+    bool noMoreBlending = true;
+    do
+    {
+      noMoreBlending
+          = tryBlend(dynamicPath, feasibilityChecker, attempt, dtShortcut);
+    } while (noMoreBlending);
```

Comment:
Nit: The name `noMoreBlending` implies the opposite of what it actually is. I suggest renaming the flag to `needsMoreBlending`.